src/util/bson/bson.hpp

```diff
@@ -32,41 +32,670 @@
 #include <ostream>
 #include <variant>
 
+#include <iostream>
 namespace realm {
 namespace bson {
 
-namespace _impl {
-/// A variant of the allowed Bson types.
-template<class T>
-using Var = std::variant<
-    Null,
-    int32_t,
-    int64_t,
-    bool,
-    float,
-    double,
-    time_t,
-    std::string,
-    std::vector<char>,
-    Timestamp,
-    Decimal128,
-    ObjectId,
-    RegularExpression,
-    MinKey,
-    MaxKey,
-    IndexedMap<T>,
-    std::vector<T>
->;
-
-// tie the knot
-template <template<class> class K>
-struct Fix : K<Fix<K>>
-{
-   using K<Fix>::K;
+class Bson {
+public:
+    Bson() noexcept
+        : m_type(NONE)
+    {
+    }
+
+    Bson(util::None) noexcept
+        : Bson()
+    {
+    }
+
+    Bson(int32_t) noexcept;
+    Bson(int64_t) noexcept;
+    Bson(bool) noexcept;
+    Bson(double) noexcept;
+    Bson(MinKey) noexcept;
+    Bson(MaxKey) noexcept;
+
+
+    Bson(time_t) noexcept;
+    Bson(Timestamp) noexcept;
+    Bson(Decimal128) noexcept;
+    Bson(ObjectId) noexcept;
+
+    Bson(const RegularExpression&) noexcept;
+    Bson(const std::vector<char>&) noexcept;
+    Bson(const std::string&) noexcept;
+    Bson(const IndexedMap<Bson>&) noexcept;
+    Bson(const std::vector<Bson>&) noexcept;
+    Bson(std::string&&) noexcept;
+    Bson(IndexedMap<Bson>&&) noexcept;
+    Bson(std::vector<Bson>&&) noexcept;
+    
+    // These are shortcuts for Bson(StringData(c_str)), and are
+    // needed to avoid unwanted implicit conversion of char* to bool.
+    Bson(char* c_str) noexcept
+        : Bson(StringData(c_str))
+    {
+    }
+    Bson(const char* c_str) noexcept
+        : Bson(StringData(c_str))
+    {
+    }
+
+    ~Bson() noexcept
+    {
+        switch (m_type) {
+            case STRING:
+                delete string_val;
+                string_val = NULL;
+                break;
+            case DOCUMENT:
+                delete document_val;
+                document_val = NULL;
+                break;
+            case ARRAY:
+                delete array_val;
+                array_val = NULL;
+                break;
+            case BINARY:
+                delete binary_val;
+                binary_val = NULL;
+                break;
+            case REGEX:
+                delete regex_val;
+                regex_val = NULL;
+                break;
+            default:
+                break;
+        }
+    }
+
+    Bson(Bson&& v) {
+        std::cout<<"TRYING TO MOVE BSON"<<std::endl;
+        m_type = v.m_type;
+    }
+
+    Bson(const Bson& v) {
```

Comment:
The copy constructor can be just `m_type = None; *this = v;` (and similar for the move constructor).